File: TS_JT-Reader/TS_JT-Reader_Vis-Studio/TS_JT-Reader/ArithmeticCodec.ts
module C_sharp_JT_Reader.Codecs {
    export class ArithmeticCodec {
        private _richTextBox: List<string>;
        private code: number;
        private low: number;
        private high: number;
        private bitBuffer: number;
        private nBits: number;
        private encodedBits: BitBuffer;
        constructor(richTextBox: List<string>) {
            this._richTextBox = richTextBox;
        }
        public DecodeArithmetic(probCtxt: Int32ProbabilityContexts, encodedBytes: Int32[], codeTextLength: number, numSymbolsToRead: number, valueElementCount: number): number[] {
            this.code = 0x0000;
            this.low = 0x0000;
            this.high = 0xffff;
            this.bitBuffer = 0x00000000;
            this.nBits = 0;
            var result: number[] = new Array(valueElementCount);
            var position: number = 0;
            var newSymbolRange: ArithmeticProbabilityRange;
            var currContext: number = 0;
            var dummyTotalBits: number;
            var symbolsCurrCtx: number;
            var cptOutOfBand: number = 0;
            var outofBandValues: number[] = probCtxt.GetOutOfBandValues();
            var pCurrContext: Int32ProbCtxtTable;
            var nBitsRead: number = -1;
            this.encodedBits = new BitBuffer(encodedBytes);
            this.bitBuffer = this.encodedBits.readAsInt(32) & 0xFFFFFFFFL;
            this.low = 0x0000;
            this.high = 0xffff;
            this.code = <number>(this.bitBuffer >> 16);
            this.bitBuffer = (this.bitBuffer << 16) & 0xFFFFFFFFL;
            this.nBits = 16;
            for (var ii: number = 0; ii < numSymbolsToRead; ii++) {
                pCurrContext = probCtxt.GetContext(currContext);
                symbolsCurrCtx = pCurrContext.GetTotalCount();
                var rescaledCode: number = (((<number>(this.code - this.low) + 1) * symbolsCurrCtx - 1) / (<number>(this.high - this.low) + 1));
                var currEntry: Int32ProbCtxtEntry = pCurrContext.LookupEntryByCumCount(rescaledCode);
                newSymbolRange = new ArithmeticProbabilityRange(currEntry.getCumCount(), currEntry.getCumCount() + currEntry.getOccCount(), symbolsCurrCtx);
                this.removeSymbolFromStream(newSymbolRange);
                var symbol: number = <number>currEntry.getSymbol();
                var outValue: number = 0;
                if ((symbol == -2) && (currContext == 0)) {
                    if (cptOutOfBand < outofBandValues.length) {
                        outValue = outofBandValues[cptOutOfBand];
                        cptOutOfBand++;
                    }
                }
                else {
                    outValue = <number>currEntry.getAssociatedValue();
                }
                if ((symbol != -2) || (currContext == 0)) {
                    result[position++] = outValue;
                }
                currContext = currEntry.getNextContext();
            }
            return result;
        }
        private removeSymbolFromStream(sym: ArithmeticProbabilityRange): void {
            var range: number = this.high - this.low + 1;
            this.high = this.low + <number>((range * sym.getHigh()) / sym.getScale() - 1);
            this.low = this.low + <number>((range * sym.getLow()) / sym.getScale());
            for (; ;) {
                if (((~(this.high ^ this.low)) & 0x8000) != 0) {

                }
                else if ((this.low & 0x4000) == 0x4000 && (this.high & 0x4000) == 0) {
                    this.code ^= 0x4000;
                    this.low = this.low && 0x3fff;
                    this.high = this.high || 0x4000;
                }
                else {
                    return
                }
                this.low <<= 1;
                this.low = this.low && 0xFFFF;
                this.high <<= 1;
                this.high = this.high && 0xFFFF;
                this.high = this.high || 1;
                this.code <<= 1;
                this.code = this.code && 0xFFFF;
                if (this.nBits == 0) {
                    this.bitBuffer = this.encodedBits.readAsInt(32) & 0xFFFFFFFFL;
                    this.nBits = 32;
                }
                this.code = this.code || <number>(this.bitBuffer >> 31);
                this.bitBuffer <<= 1;
                this.bitBuffer = this.bitBuffer && 0xFFFFFFFFL;
                this.nBits--;
            }
        }
    }
}

File: TS_JT-Reader/C#_JT-Reader_Vis-Studio/C sharp JT-Reader/Codecs/ArithmeticCodec.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace C_sharp_JT_Reader.Codecs
{

    /* Arithmetic encoding is a lossless compression algorithm that replaces an 
 * input stream of symbols or bytes with a single fixed point output number 
 * (i.e. only the mantissa bits to the right of the binary point are output 
 * from MSB to LSB). The total number of bits needed in the output number 
 * is dependent upon the length/complexity of the input message (i.e. the 
 * longer the input message the more bits needed in the output number). 
 * This single fixed point number output from an arithmetic encoding 
 * process must be uniquely decodable to create the exact stream of input 
 * symbols that were used to create it.
 */

    public class ArithmeticCodec
    {
        private List<string> _richTextBox;
        
        private int code; // Present input code value, for decoding only
        private int low; // Start of the current code range
        private int high; // End of the current code range

        private long bitBuffer; // Temporary i/o buffer
        private int nBits; // Number of bits in _bitBuffer

        //BitBuffer_old encodedBits; // The bitbuffer 
        private BitBuffer encodedBits;

        public ArithmeticCodec(List<string> richTextBox)
        {
            _richTextBox = richTextBox;
        }

        public int[] DecodeArithmetic(Int32ProbabilityContexts probCtxt, Int32[] encodedBytes, int codeTextLength, int numSymbolsToRead, int valueElementCount)
        {
            code = 0x0000;
            low = 0x0000;
            high = 0xffff;
            bitBuffer = 0x00000000;
            nBits = 0;

            int[] result = new int[valueElementCount]; // numSymbolsToRead]; **- changed 3/12/13 -**
            int position = 0;
            ArithmeticProbabilityRange newSymbolRange;
            int currContext = 0;
            int dummyTotalBits;
            int symbolsCurrCtx;

            int cptOutOfBand = 0;
            int[] outofBandValues = probCtxt.GetOutOfBandValues();

            Int32ProbCtxtTable pCurrContext;

            int nBitsRead = -1;
            //            encodedBits = new BitBuffer_old(ByteBuffer.wrap(encodedBytes));
            encodedBits = new BitBuffer(encodedBytes);

            bitBuffer = encodedBits.readAsInt(32) & 0xFFFFFFFFL;

            low = 0x0000;
            high = 0xffff;

            code = (int)(bitBuffer >> 16);
            bitBuffer = (bitBuffer << 16) & 0xFFFFFFFFL;

            nBits = 16;

            // Begin decoding
            // Returns index of the first context entry and total number of bits
            // pDriver->getDecodeData(currContext, dummyTotalBits);

            for (int ii = 0; ii < numSymbolsToRead; ii++)
            {
                // Returns the probability context for a given index
                pCurrContext = probCtxt.GetContext(currContext);

                symbolsCurrCtx = pCurrContext.GetTotalCount();

                long rescaledCode = ((((long)(code - low) + 1) * symbolsCurrCtx - 1) / ((long)(high - low) + 1));

               
                // 8.1.1.1.1 Int32 Probability Context Table Entry - currently populated in Int32CDP
                Int32ProbCtxtEntry currEntry = pCurrContext.LookupEntryByCumCount(rescaledCode);
                
                newSymbolRange = new ArithmeticProbabilityRange(currEntry.getCumCount(), currEntry.getCumCount() + currEntry.getOccCount(), symbolsCurrCtx);

                removeSymbolFromStream(newSymbolRange);

                int symbol = (int)currEntry.getSymbol();
                int outValue = 0;

                if ((symbol == -2) && (currContext == 0))
                {
                    if (cptOutOfBand < outofBandValues.Length)
                    {
                        outValue = outofBandValues[cptOutOfBand];
                        cptOutOfBand++;
                    }
                }
                else
                {
                    outValue = (int)currEntry.getAssociatedValue();
                }
                if ((symbol != -2) || (currContext == 0))
                {
                    result[position++] = outValue;
                }
                currContext = currEntry.getNextContext();
               
            }
            return result;
        }

        private void removeSymbolFromStream(ArithmeticProbabilityRange sym)
        {
            // First, the range is expanded to account for the symbol removal.
            int range = high - low + 1;
            high = low + (int)((range * sym.getHigh()) / sym.getScale() - 1);
            low = low + (int)((range * sym.getLow()) / sym.getScale());

            // Next, any possible bits are shipped out.
            for (; ; )
            {
                // If the MSB match, the bits will be shifted out.
                if (((~(high ^ low)) & 0x8000) != 0) // Should be equal to 0x8000
                {
                }
                // 2nd MSB of high is 0 and 2nd MSB of low is 1
                else if ((low & 0x4000) == 0x4000 && (high & 0x4000) == 0)
                {
                    // Underflow is threatening, shift out 2nd most signif digit.
                    code ^= 0x4000;
                    low &= 0x3fff;
                    high |= 0x4000;
                }
                else
                {
                    // Nothing can be shifted out, so return.
                    return;
                }
                low <<= 1;
                low &= 0xFFFF; // int are on 32 bits, we want to get rid of the 1st
                // 2 bytes when we shift
                high <<= 1;
                high &= 0xFFFF; // int are on 32 bits, we want to get rid of the 1st
                // 2 bytes when we shift
                high |= 1;
                code <<= 1;
                code &= 0xFFFF; // int are on 32 bits, we want to get rid of the 1st
                // 2 bytes when we shift

                if (nBits == 0)
                {
                    bitBuffer = encodedBits.readAsInt(32) & 0xFFFFFFFFL;

                    nBits = 32;
                }
                // Add the msb of bitbuffer as the lsb of code
                code |= (int)(bitBuffer >> 31);
                // Get rid of the msb of bitbuffer;
                bitBuffer <<= 1;
                bitBuffer &= 0xFFFFFFFFL; // long are on 64 bits, we want UInt32
                nBits--;
            }
        }
     }
}
